src/rascal/representations/calculator_spherical_expansion.hh

```diff
@@ -1086,31 +1124,389 @@ namespace rascal {
+                          << " with species=0, angular_l=0 and max_radial=0"
+                          << json_projection_matrices.at(species)
+                                 .at(angular_l)
+                                 .at(radial_c)
+                                 .size()
+                          << ".";
+                  throw std::logic_error(err_str.str());
+                }
+              }
+            }
+          }
+
+          // init projection matrices
+          Matrix_t projection_matrix{};
+          std::vector<Matrix_t> angular_projection_matrices(this->max_angular +
+                                                            1);
+          for (auto it = json_projection_matrices.begin();
+               it != json_projection_matrices.end(); ++it) {
+            species = it->first;
+            for (size_t angular_l = 0; angular_l < this->max_angular + 1;
+                 angular_l++) {
+              projection_matrix.resize(this->n_components, this->max_radial);
+              projection_matrix.setZero();
+              for (size_t radial_c = 0; radial_c < this->n_components;
+                   radial_c++) {
+                for (size_t radial_n = 0; radial_n < this->max_radial;
+                     radial_n++) {
+                  projection_matrix(radial_c, radial_n) =
+                      json_projection_matrices.at(species)
+                          .at(angular_l)
+                          .at(radial_c)
+                          .at(radial_n);
+                }
+              }
+              angular_projection_matrices.at(angular_l) = projection_matrix;
+            }
+            this->projection_matrices.insert(
+                std::pair<int, std::vector<Matrix_t>>(
+                    std::stoi(species), angular_projection_matrices));
+          }
+
+          this->init_matrices();
+        } else {
+          std::stringstream err_str{};
+          err_str << "No projection matrices were given.";
+          throw std::logic_error(err_str.str());
+        }
+        Parent::init_matrices();
+      }
+
+      void precompute() override {
+        Parent::precompute();
+        this->precompute_fac_a();
+        this->precompute_center_contribution();
+      }
+
+      void init_matrices() override {
+        this->reduced_radial_integral_neighbour.resize(this->n_components,
+                                                       this->max_angular + 1);
+        this->reduced_radial_integral_neighbour.setZero();
+        this->reduced_radial_neighbour_derivative.resize(this->n_components,
+                                                         this->max_angular + 1);
+        this->reduced_radial_neighbour_derivative.setZero();
+      }
+
+      // Returns the precomputed center contribution
+      template <size_t Order, size_t Layer>
+      Vector_Ref
+      compute_center_contribution(ClusterRefKey<Order, Layer> & /*center*/,
+                                  int center_type) {
```

Comment:
I just realized `compute_neighbour_contribution` and `compute_center_contribution` with `RadialDimReductionSpline` now requires the center type as input. This change of the signature is unnecessary since you have `center` at your disposable to get this information in the function. Moreover it these functions are expected to have a particular signature so to make this can of changes one would have to change the signature of every other `compute_center_contribution` and `compute_neighbour_contribution`.
Please remove the atom type input and get it from `center` and `neighbor`.

Reply:
I did this because in the level of `compute_center_contribution` it is StructureManager where i can use `get_atom_type`, but within ` compute_[center,neighbour]_contribution` it is upcasted to ClusterRefKey and has no `get_atom_type`

Reply:
you are right, forgot about this difference. Probably `ClusterRefKey` should also have the atom type but that's another story.